community/learn/graphql-tutorials/tutorials/react-apollo-hooks/app-final/src/components/Todo/TodoInput.js

```diff
@@ -0,0 +1,68 @@
+import React, {useState} from 'react';
+import { useMutation } from '@apollo/react-hooks';
+import gql from 'graphql-tag';
+import {GET_MY_TODOS} from './TodoPrivateList';
+
+const ADD_TODO = gql `
+  mutation ($todo: String!, $isPublic: Boolean!) {
+    insert_todos(objects: {title: $todo, is_public: $isPublic}) {
+      affected_rows
+      returning {
+        id
+        title
+        created_at
+        is_completed
+      }
+    }
+  }
+`;
+
+const TodoInput = ({isPublic = false}) => {
+  let input;
+
+  const [todoInput, setTodoInput] = useState('');
+
+  const updateCache = (cache, {data}) => {
+      // If this is for the public feed, do nothing
+    if (isPublic) {
+      return null;
+    }
+
+    // Fetch the todos from the cache
+    const existingTodos = cache.readQuery({
+      query: GET_MY_TODOS
+    });
+
+    // Add the new todo to the cache
+    const newTodo = data.insert_todos.returning[0];
+    cache.writeQuery({
+      query: GET_MY_TODOS,
+      data: {todos: [newTodo, ...existingTodos.todos]}
+    });
+  };
+
+  const resetInput = () => {
+    setTodoInput('');
+    input.focus();
+  };
+
+  const [addTodo, {loading}] = useMutation(ADD_TODO);
+
+  return (
+    <form className="formInput" onSubmit={(e)=> {
+      e.preventDefault();
+      addTodo({variables: {todo: todoInput, isPublic }, update: updateCache, onCompleted: resetInput()});
```

Comment:
Yes I have moved update and onCompleted in the useMutation options, does make it cleaner. But input.focus() won't work now since input is null, don't know how to get around it. Let me know what should be done.

Reply:
I think it should work, except, resetInput would be called "after" the mutation is complete and not immediately after you submit.

If it is not working, can you tell the exact reason why?

Reply:
I can't exactly say but the input variable is null, when resetInput is called.

Reply:
Removed the disabled attribute and field gets its focus. Not sure, why the input field is null in that. I checked the input variable has value in the onSubmit function but becomes null in resetInput.